File: datamodel/src/main/java/datamodel/clinical/treatment/Radiotherapy.java
package datamodel.clinical.treatment;

import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class Radiotherapy implements Therapy {
    public final TreatmentType treatmentType = TreatmentType.RADIOTHERAPY;

    @Override
    @NotNull
    @Value.Default
    public Set<Drug> drugs() {
        return Collections.emptySet();
    }

    @Override
    @NotNull
    public Set<TreatmentCategory> categories() {
        return Stream.concat(Stream.of(TreatmentCategory.RADIOTHERAPY), drugs().stream().map(Drug::category)).collect(Collectors.toSet());
    }

    @Nullable
    public abstract String radioType();

    @Nullable
    public abstract Boolean isInternal();
}

File: algo/src/main/kotlin/com/hartwig/actin/clinical/datamodel/treatment/Radiotherapy.kt
package com.hartwig.actin.clinical.datamodel.treatment

data class Radiotherapy(
    override val name: String,
    override val synonyms: Set<String> = emptySet(),
    override val displayOverride: String? = null,
    override val isSystemic: Boolean = false,
    val radioType: RadiotherapyType? = null,
    val isInternal: Boolean? = null
) : Treatment {

    override val treatmentClass = TreatmentClass.RADIOTHERAPY

    override fun types() = setOfNotNull(radioType)
   
    override fun categories() = setOf(TreatmentCategory.RADIOTHERAPY)
}
